master/internal/db/postgres.go

```diff
@@ -1441,6 +1452,211 @@ func (db *PgDB) UpdateStep(
 	if err != nil {
 		return errors.Wrapf(err, "error finding step (%v, %v) to update", trialID, totalBatches)
 	}
+	return db.updateStep(step, newState, metrics)
+}
+
+// AddTrainingMetrics adds a completed step to the database with the given training metrics.
+// If these training metrics occur before any others, a rollback is assumed and later
+// training and validation metrics are cleaned up.
+func (db *PgDB) AddTrainingMetrics(ctx context.Context, m *trialv1.TrainingMetrics) error {
+	return db.withTransaction("add training metrics", func(tx *sqlx.Tx) error {
+		var cRunID int
+		switch err := tx.QueryRowxContext(ctx, `
+SELECT id
+FROM trial_runs
+WHERE trial_id = $1
+`, m.TrialId).Scan(&cRunID); {
+		case err != nil:
+			return errors.Wrap(err, "querying current run")
+		case int(m.TrialRunId) < cRunID:
+			return api.AsErrBadRequest("stale training metrics for run %d during %d", m.TrialRunId, cRunID)
+		}
+
+		if _, err := tx.ExecContext(ctx, `
+UPDATE raw_steps SET archived = true
+WHERE trial_id = $1
+  AND trial_run_id < $2
+  AND total_batches >= $3;
+`, m.TrialId, m.TrialRunId, m.BatchesTrained); err != nil {
+			return errors.Wrap(err, "archiving training metrics")
+		}
+
+		if _, err := tx.ExecContext(ctx, `
+UPDATE raw_validations SET archived = true
+WHERE trial_id = $1
+  AND trial_run_id < $2
+  AND total_batches > $3;
+`, m.TrialId, m.TrialRunId, m.BatchesTrained); err != nil {
+			return errors.Wrap(err, "archiving validations")
+		}
+
+		// TODO(DET-5210): This can go away when step ID does.
+		var id int
+		if err := tx.QueryRowxContext(ctx, `
+SELECT max(id)
+FROM steps
+WHERE trial_id = $1`, m.TrialId).Scan(&id); err != nil {
+			return errors.Wrap(err, "querying next id")
+		}
+
+		startTime, err := derivePriorWorkloadEndTime(ctx, tx, int(m.TrialId))
```

Comment:
replied in thread on slack, I agree, I think now the SDK should handle deriving it

Reply:
resolution on this (from slack) was that we agree that `start_time` shouldn't be calculated from step start/end times, and it will be easy to recalculate it on new information at another time.  The only real reason to keep start_time around was for resource allocation, which read the steps table, but that information can be more correctly calculated from the new runs table.

Also, by leaving the derived time here (which is correct in the current system, albeit fragile) we can land this work without having to rewrite resource allocation logic before landing this PR.